src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,369 @@
+SG_PRIMITIVE_TYPE(SGMatrix<float64_t>, TYPE::PT_SGMATRIX_FLOAT64)
+SG_PRIMITIVE_TYPE(SGMatrix<floatmax_t>, TYPE::PT_SGMATRIX_FLOATMAX)
+SG_PRIMITIVE_TYPE(SGMatrix<int32_t>, TYPE::PT_SGMATRIX_INT32)
+SG_PRIMITIVE_TYPE(SGMatrix<int64_t>, TYPE::PT_SGMATRIX_INT64)
+
+#undef SG_PRIMITIVE_TYPE
+
+namespace shogun
+{
+	typedef std::unordered_map<std::type_index, TYPE> typemap;
+
+	namespace type_internal
+	{
+		std::string print_map(const typemap& map)
+		{
+			auto msg = std::string("<");
+			for (auto it : map)
+			{
+				msg += demangled_type(it.first.name());
+				msg += ", ";
+			}
+			msg.resize(msg.size() - 2);
+			msg += ">";
+			return msg;
+		}
+
+		TYPE get_type(const Any& any, const typemap& map)
+		{
+			auto type = std::type_index(any.type_info());
+			auto it = map.find(type);
+
+			return it == map.end() ? TYPE::PT_UNDEFINED : map.at(type);
+		}
+
+		template <typename T>
+		struct is_vector : public std::false_type
+		{
+		};
+		template <typename T>
+		struct is_vector<SGVector<T>> : public std::true_type
+		{
+		};
+
+		template <typename T>
+		struct is_matrix : public std::false_type
+		{
+		};
+		template <typename T>
+		struct is_matrix<SGMatrix<T>> : public std::true_type
+		{
+		};
+
+		template <
+			typename T, typename ScalarLambdaT, typename VectorLambdaT,
+			typename MatrixLambdaT>
+		typename std::enable_if_t<
+			std::is_scalar<T>::value and
+				not std::is_same<ScalarLambdaT, std::nullptr_t>::value,
+			void>
+		execute_function(
+			T value, ScalarLambdaT scalar_func, VectorLambdaT vector_func,
+			MatrixLambdaT matrix_func)
+		{
+			scalar_func(value);
+		}
+
+		template <
+			typename T, typename ScalarLambdaT, typename VectorLambdaT,
+			typename MatrixLambdaT>
+		typename std::enable_if_t<
+			is_vector<T>::value and
+				not std::is_same<VectorLambdaT, std::nullptr_t>::value,
+			void>
+		execute_function(
+			T value, ScalarLambdaT scalar_func, VectorLambdaT vector_func,
+			MatrixLambdaT matrix_func)
+		{
+			vector_func(value);
+		}
+
+		template <
+			typename T, typename ScalarLambdaT, typename VectorLambdaT,
+			typename MatrixLambdaT>
+		typename std::enable_if_t<
+			is_matrix<T>::value and
+				not std::is_same<MatrixLambdaT, std::nullptr_t>::value,
+			void>
+		execute_function(
+			T value, ScalarLambdaT scalar_func, VectorLambdaT vector_func,
+			MatrixLambdaT matrix_func)
+		{
+			matrix_func(value);
+		}
+
+		template <
+			typename T, typename ScalarLambdaT, typename VectorLambdaT,
+			typename MatrixLambdaT>
+		typename std::enable_if_t<
+			(std::is_same<ScalarLambdaT, std::nullptr_t>::value and
+			 std::is_scalar<T>::value) or
+				(std::is_same<VectorLambdaT, std::nullptr_t>::value and
+				 is_vector<T>::value) or
+				(std::is_same<MatrixLambdaT, std::nullptr_t>::value and
+				 is_matrix<T>::value),
+			void>
+		execute_function(
+			T value, ScalarLambdaT scalar_func, VectorLambdaT vector_func,
+			MatrixLambdaT matrix_func)
+		{
+			SG_SWARNING(
```

Comment:
I feel this should not be a runtime warning?

Reply:
I have been thinking that a lot of these things should be static asserts to make debugging easier... If everything compiles any developer might become more complacent...

Reply:
We could check the signature of the function to see if it is void and has one parameter at compile time.

Reply:
I like such static assertions a lot!

Reply:
Alright, got 1 static assertion working! I didn't work on anything else yet as this took a while to figure out. I also wanted to add a test for static_assert with gtest, and that wasn't easy but now its working! :D

Reply:
Should I replace the warning with an exception? The idea of the warning was that there could be situations where no lambda implementation exists of the expected type on purpose and an exception could abort a ML run without need.

Reply:
> Alright, got 1 static assertion working! I didn't work on anything else yet as this took a while to figure out. I also wanted to add a test for static_assert with gtest, and that wasn't easy but now its working! :D

nice one!

Reply:
> Should I replace the warning with an exception? The idea of the warning was that there could be situations where no lambda implementation exists of the expected type on purpose and an exception could abort a ML run without need.

I would say yes, with an appropriate string describing what is wrong